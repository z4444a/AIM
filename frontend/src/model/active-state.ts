export const ActiveState = [
  {
    key: '1',
    value: 'Активен',
  },
  {
    key: '2',
    value: 'Не активен',
  },
];
