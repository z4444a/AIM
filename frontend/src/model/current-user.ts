import { Role } from '../commons/role';

export interface CurrentUser {
  id: number;
  fullName: string;
  username: string;
  role: Role;
}
