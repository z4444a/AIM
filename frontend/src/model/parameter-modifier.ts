export enum ParameterModifier {
  POOL_PARAMETER = 1,
  REQUEST_PARAMETER = 2,
  ALLOCATION_PARAMETER = 3,
}
