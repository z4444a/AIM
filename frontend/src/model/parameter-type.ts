export enum ParameterType {
  NUMBER = 'NUMBER',
  TEXT = 'TEXT',
  DATE = 'DATE',
  LIST = 'LIST',
  REAL = 'REAL',
  POOL = 'POOL',
}
