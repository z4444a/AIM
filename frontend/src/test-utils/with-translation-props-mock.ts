import { WithTranslation } from 'react-i18next';

export function getMockWithTranslationProps(): WithTranslation {
  // eslint-disable-next-line
  const t = ((key: string) => `${key}_i18n`) as any;
  // eslint-disable-next-line
  const i18n = {} as any;
  const props: WithTranslation = {
    tReady: true,
    t,
    i18n,
  };

  return props;
}
