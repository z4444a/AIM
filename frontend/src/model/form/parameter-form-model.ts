import { ParameterConstraintFormModel } from './parameter-constraint-form-model';
import { ParameterType } from '../parameter-type';
import { ParameterModifier } from '../parameter-modifier';

export interface ParameterFormModel {
  id?: number;
  key?: number;
  name: string;
  identifier: string;
  required: boolean;
  visibleToOwner?: boolean;
  modifier: ParameterModifier;
  parameterType: ParameterType;
  constraint: ParameterConstraintFormModel;
  order: number;
  poolTypeId?: number;
}
