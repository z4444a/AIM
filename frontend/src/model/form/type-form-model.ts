import { ParameterFormModel } from './parameter-form-model';
import { PictureFormModel } from './picture-form-model';

export interface TypeFormModel {
  id?: number;
  name: string;
  description: string;
  active: boolean;
  needsBackup: boolean;
  quantitative: boolean;
  picture: PictureFormModel;
  parameters: ParameterFormModel[];
}
