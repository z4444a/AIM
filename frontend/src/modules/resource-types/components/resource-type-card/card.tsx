import React, { Component } from 'react';
import { Style } from './styles';
import { WithTranslation } from 'react-i18next';
import { Card } from '@material-ui/core';
import Divider from '@material-ui/core/Divider';
import Button from '@material-ui/core/Button';
import TypeFields from './fields/index';
import { TypeFormModel } from '../../../../model/form/type-form-model';
import { ParameterFormModel } from '../../../../model/form/parameter-form-model';
import ParameterList from './parameters/parameter-list/index';
import { Props as ParameterListProps } from './parameters/parameter-list/parameter-list';
import { ParameterRowFormModel } from '../../../../model/form/parameter-row-form-model';
import _ from 'underscore';
import { ParameterModifier } from '../../../../model/parameter-modifier';
import Tooltip from '@material-ui/core/Tooltip';
import { ResourceTypeSuggestion } from '../../../../redux/reducers/request-create-page-reducer';
import { ValidationParameterResponse } from '../../../../model/validation-parameter-response';
import { ParameterUniqueFields } from '../../../../model/parameters/parameter-unique-fields';

export interface Props {
  dialogIsOpen: boolean;
  type: TypeFormModel;
  uniqueName: boolean;
  parameterRows: ParameterRowFormModel[];
  highlightEmpty: boolean;
  showIdentifiers?: boolean;

  onAddParameterRow: (modifier: ParameterModifier) => void;
  onChange: (type: TypeFormModel) => void;
  onNameValidationRequested: (name: string) => boolean;
  onDialogOpened: () => void;
  onDialogClosed: () => void;
  onParameterEditSet: (name: string) => void;
  onParameterEditReset: () => void;
  onParameterFormChange: (index: number, model: ParameterFormModel) => void;
  onHighlightEmptySet: () => void;
  onHighlightEmptyReset: () => void;
  onCompleteButtonClick: () => void;
  onCancelButtonClick: () => void;
  onParameterRowEditSet: (key: number, editable: boolean) => void;
  onParameterRowRemoval: (key: number) => void;
  editedParameterName: string | null;
  changeOrder: (rows: ParameterRowFormModel[]) => void;

  onShowOverlay: (text: string) => void;
  onHideOverlay: () => void;

  findParamValue?: (id: number) => void;
  removedParameterResponse: ValidationParameterResponse | null;
  resetRemoveParameterResponse?: () => void;

  typeSuggestions: ResourceTypeSuggestion[];

  findParamIdentifier: (params: ParameterUniqueFields) => void;
  identifierValidation: ValidationParameterResponse | null;
  setValidation: (params: ValidationParameterResponse | null) => void;
}

export type InternalProps = Props & Style & WithTranslation;

export interface ErrorState {
  picture?: string;
  showTooltip: boolean;
}

export class ResourceTypesCard extends Component<InternalProps, ErrorState> {
  constructor(props: InternalProps) {
    super(props);
    this.state = {
      showTooltip: true,
    };
  }
  public render(): JSX.Element {
    const {
      classes,
      type,
      uniqueName,
      parameterRows,
      onCancelButtonClick,
      onParameterFormChange,
      onParameterRowEditSet,
      highlightEmpty,
      onAddParameterRow,
      t,

      onShowOverlay,
      onHideOverlay,

      findParamValue,
      removedParameterResponse,
      resetRemoveParameterResponse,

      typeSuggestions,

      findParamIdentifier,
      identifierValidation,
      setValidation,
    } = this.props;
    const { name, description, active, needsBackup, picture, quantitative } = type;
    const parameterListProps: ParameterListProps = {
      onParameterRowModelChange: onParameterFormChange,
      onRemoveButtonClick: this.handleParameterRowRemoval,
      onRowEditSet: onParameterRowEditSet,
      onEditConfirmed: this.handleParameterEditConfirmed,
      onEditCanceled: this.handleParameterEditReset,
      onAddParameterRow: onAddParameterRow,
      highlightEmpty: highlightEmpty,
      rows: parameterRows,
      changeOrder: this.handleChangeOrder,

      onShowOverlay: onShowOverlay,
      onHideOverlay: onHideOverlay,

      findParamValue: findParamValue,
      removedParameterResponse: removedParameterResponse,
      resetRemoveParameterResponse: resetRemoveParameterResponse,

      typeSuggestions: typeSuggestions,

      findParamIdentifier: findParamIdentifier,
      identifierValidation: identifierValidation,
      setValidation: setValidation,
    };
    return (
      <Card className={classes.container}>
        <TypeFields
          errors={this.state}
          imgSource={picture.picturePath}
          imgBytes={picture.picture}
          name={name}
          uniqueName={uniqueName}
          description={description}
          active={active}
          backup={needsBackup}
          quantitative={quantitative}
          onImgByteCodeChange={this.handlePictureIconChange}
          onImgSourceChange={this.handlePictureAddressChange}
          onNameChange={this.handleNameChange}
          onDescriptionChange={this.handleDescriptionChange}
          onActiveChange={this.handleCheckBoxChange('active')}
          onBackupChange={this.handleCheckBoxChange('backup')}
          onQuantitativeChange={this.handleCheckBoxChange('quantitative')}
          highlightEmpty={highlightEmpty}
        />
        <Divider />
        <ParameterList {...parameterListProps} />
        <div className={classes.buttonGroup}>
          <Button variant="contained" onClick={onCancelButtonClick}>
            {t('common.cancel')}
          </Button>
          <Tooltip title={this.state.showTooltip ? t('typeCard.tooltips.hasUnconfirmedParam') : ''}>
            <Button
              className={classes.completeButton}
              variant="contained"
              onClick={this.handleCompletion}
            >
              {t('common.complete')}
            </Button>
          </Tooltip>
        </div>
      </Card>
    );
  }

  private handlePictureIconChange = (bytes: number[]) => {
    const { onChange, type } = this.props;
    const newType = {
      ...type,
      picture: {
        ...type.picture,
        picturePath: '',
        picture: bytes,
      },
    };

    onChange(newType);
    this.setState({
      picture: undefined,
    });
  };

  private handlePictureAddressChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>
  ) => {
    const { onChange, type } = this.props;

    const newType = {
      ...type,
      picture: {
        ...type.picture,
        picturePath: e.target.value,
        picture: [],
      },
    };

    onChange(newType);
    this.setState({
      picture: undefined,
    });
  };

  private handleNameChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>
  ) => {
    const { onChange, type, onNameValidationRequested } = this.props;

    const newType = {
      ...type,
      name: e.target.value,
    };

    onChange(newType);
    onNameValidationRequested(e.target.value);
  };

  private handleCheckBoxChange = (field: 'active' | 'backup' | 'quantitative') => (
    e: React.ChangeEvent<HTMLInputElement>,
    checked: boolean
  ) => {
    const { onChange, type } = this.props;

    let active = type.active;
    let backup = type.needsBackup;
    let quantitative = type.quantitative;

    switch (field) {
      case 'active':
        active = checked;
        break;
      case 'backup':
        backup = checked;
        break;
      case 'quantitative':
        quantitative = checked;
    }

    const newType = {
      ...type,
      active: active,
      needsBackup: backup,
      quantitative: quantitative,
    };

    onChange(newType);
  };

  private handleDescriptionChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>
  ) => {
    const { onChange, type } = this.props;

    const newType = {
      ...type,
      description: e.target.value,
    };

    onChange(newType);
  };

  private handleParameterEditConfirmed = (key: number) => {
    const { type, parameterRows, onChange, onParameterRowEditSet } = this.props;
    const { parameters } = type;
    const row = parameterRows.find(row => row.key === key);
    if (!row) {
      return;
    }

    const newParameter = {
      key: row.key,
      ...row.model,
    };

    const changedIndex = parameters.findIndex(value => value.key === newParameter.key);

    const newParams = _.clone(parameters);
    if (changedIndex > -1) {
      newParams[changedIndex] = newParameter;
    } else {
      newParams.push(newParameter);
    }

    const newType = {
      ...type,
      parameters: newParams,
    };
    onChange(newType);
    onParameterRowEditSet(key, false);

    this.setState({
      showTooltip: false,
    });
  };

  private handleParameterEditReset = (key: number) => {
    const {
      type,
      parameterRows,
      onParameterFormChange,
      onParameterRowRemoval,
      onParameterRowEditSet,
    } = this.props;
    const { parameters } = type;
    const row = parameterRows.find(row => row.key === key);
    if (!row) {
      return;
    }

    const changedIndex = parameters.findIndex(value => value.key === row.key);

    if (changedIndex > -1) {
      const prevValue = parameters[changedIndex];
      onParameterFormChange(key, prevValue);
      onParameterRowEditSet(key, false);
    } else {
      onParameterRowRemoval(key);
    }
  };

  private validateType = () => {
    const { name } = this.props.type;
    const picture = this.validatePicture();
    return !!name && this.props.uniqueName && picture;
  };

  private validatePicture = () => {
    const { type, t } = this.props;
    const { picturePath } = type.picture;
    if (picturePath === '') {
      return true;
    }
    const match = picturePath.match(new RegExp('^https?://'));
    const correctPicture = match !== null;
    if (!correctPicture) {
      this.setState({
        picture: t('typeCard.fields.icon.shouldBeginWithHttps'),
      });
    }
    return correctPicture;
  };

  private handleCompletion = () => {
    const { onHighlightEmptySet, parameterRows, onCompleteButtonClick } = this.props;
    const error = !this.validateType() || parameterRows.some(value => value.editable);
    this.setState({
      showTooltip: error,
    });
    if (error) {
      onHighlightEmptySet();
      return;
    }
    onCompleteButtonClick();
  };

  private handleParameterRowRemoval = (key: number) => {
    const { onChange, onParameterRowRemoval, type } = this.props;
    const { parameters } = type;
    onParameterRowRemoval(key);
    const deletedIndex = parameters.findIndex(value => value.key === key);
    if (deletedIndex !== -1) {
      const newParams = [...parameters];
      newParams.splice(deletedIndex, 1);
      const refreshOrder = (item: ParameterFormModel, index: number): ParameterFormModel => {
        const { order, ...rest } = item;
        return {
          ...rest,
          order: index,
        };
      };
      const newType = {
        ...type,
        parameters: newParams.sort((a, b) => a.order - b.order).map(refreshOrder),
      };
      onChange(newType);
    }
  };
  private handleChangeOrder = (modifier: ParameterModifier, rows: ParameterRowFormModel[]) => {
    const { changeOrder, parameterRows } = this.props;
    const oldRows = parameterRows.filter(row => row.model.modifier !== modifier);
    const newRows = [...oldRows, ...rows];
    changeOrder(newRows);
  };
}
