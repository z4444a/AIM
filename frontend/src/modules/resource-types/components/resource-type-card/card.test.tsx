import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { InternalProps, ResourceTypesCard } from './card';
import { TypeFormModel } from '../../../../model/form/type-form-model';
import { ParameterFormModel } from '../../../../model/form/parameter-form-model';
import { ParameterType } from '../../../../model/parameter-type';
import { getMockWithTranslationProps } from '../../../../test-utils/with-translation-props-mock';
import { ParameterModifier } from '../../../../model/parameter-modifier';

describe('ResourceTypesCard', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const type: TypeFormModel = {
      active: true,
      description: '',
      id: 0,
      name: '',
      needsBackup: false,
      parameters: [],
      picture: {
        picture: [],
        picturePath: '',
      },
      quantitative: false,
    };

    const parameterFormModel: ParameterFormModel = {
      constraint: {},
      identifier: '',
      name: '',
      parameterType: ParameterType.NUMBER,
      required: false,
      modifier: ParameterModifier.REQUEST_PARAMETER,
    };

    const props: InternalProps = {
      ...getMockWithTranslationProps(),
      classes: {
        buttonGroup: 'buttonGroup',
        completeButton: 'completeButton',
        container: 'container',
      },
      dialogIsOpen: false,
      editedParameterName: null,
      highlightEmpty: false,
      onCancelButtonClick: () => null,
      onChange: (p1: TypeFormModel) => null,
      onCompleteButtonClick: () => null,
      onDialogClosed: () => null,
      onDialogOpened: () => null,
      onHighlightEmptyReset: () => null,
      onHighlightEmptySet: () => null,
      onParameterEditReset: () => null,
      onParameterEditSet: () => null,
      onParameterFormChange: () => null,
      parameterRows: parameterFormModel,
      type: type,
    };
    wrapper = shallow(<ResourceTypesCard {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
