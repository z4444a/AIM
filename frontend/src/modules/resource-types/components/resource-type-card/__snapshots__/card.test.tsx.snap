// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTypesCard should render correctly 1`] = `
<WithStyles(Card)
  className="container"
>
  <withI18nextTranslation(WithStyles(TypeFields))
    active={true}
    backup={false}
    description=""
    errors={
      Object {
        "showTooltip": true,
      }
    }
    highlightEmpty={false}
    imgBytes={Array []}
    imgSource=""
    name=""
    onActiveChange={[Function]}
    onBackupChange={[Function]}
    onDescriptionChange={[Function]}
    onImgByteCodeChange={[Function]}
    onImgSourceChange={[Function]}
    onNameChange={[Function]}
    onQuantitativeChange={[Function]}
    quantitative={false}
  />
  <WithStyles(Divider) />
  <withI18nextTranslation(WithStyles(ParameterList))
    changeOrder={[Function]}
    highlightEmpty={false}
    onEditCanceled={[Function]}
    onEditConfirmed={[Function]}
    onParameterRowModelChange={[Function]}
    onRemoveButtonClick={[Function]}
    rows={
      Object {
        "constraint": Object {},
        "identifier": "",
        "modifier": 2,
        "name": "",
        "parameterType": "NUMBER",
        "required": false,
      }
    }
  />
  <div
    className="buttonGroup"
  >
    <WithStyles(Button)
      onClick={[Function]}
      variant="contained"
    >
      common.cancel_i18n
    </WithStyles(Button)>
    <WithStyles(Tooltip)
      title="typeCard.tooltips.hasUnconfirmedParam_i18n"
    >
      <WithStyles(Button)
        className="completeButton"
        onClick={[Function]}
        variant="contained"
      >
        common.complete_i18n
      </WithStyles(Button)>
    </WithStyles(Tooltip)>
  </div>
</WithStyles(Card)>
`;
